File: python-learning-path/practise/fibonacci.py
"""
Exercise 02: Fibonacci Sequence

CONCEPT:
Each number is the sum of the two preceding ones.
Sequence: 0, 1, 1, 2, 3, 5, 8, ...
"""

def generate_fibonacci(n):
    if n <= 0: return []
    if n == 1: return [0]
    
    # Initialize list with first two terms
    fib = [0, 1]
    
    # Loop to generate terms from index 2 to n-1
    for i in range(2, n):
        # Add the sum of the last two elements
        next_val = fib[-1] + fib[-2]
        fib.append(next_val)
        
    return fib

def main():
    try:
        n = int(input("How many Fibonacci terms to generate? ") or "0")
        
        result = generate_fibonacci(n)
        
        # Format the list into a string for pretty printing
        print(f"Sequence: {', '.join(map(str, result))}")
        
    except ValueError:
        print("Invalid input. Please enter an integer.")

if __name__ == "__main__":
    main()

File: java-learning-path/practise/Fibonacci.java
import java.util.Scanner;

/**
 * Exercise 02: Fibonacci Sequence
 * 
 * ALGORITHM:
 * Next term = Sum of the previous two terms.
 * f(1) = 0, f(2) = 1, f(3) = (0+1) = 1, f(4) = (1+1) = 2, ...
 */
public class Fibonacci {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the number of terms: ");
        int n = scanner.nextInt();

        int first = 0, second = 1;
        System.out.print("Fibonacci Sequence up to " + n + " terms: ");

        // Loop to generate and print terms
        for (int i = 1; i <= n; ++i) {
            System.out.print(first + (i == n ? "" : ", "));
            
            // Calculate the next term
            int next = first + second;
            
            // Update pointers: 'first' moves to 'second', 'second' moves to 'next'
            first = second;
            second = next;
        }
        System.out.println();
        scanner.close();
    }
}
